pyo3-macros-backend/src/method.rs

```diff
@@ -115,146 +117,188 @@ pub fn parse_method_receiver(arg: &syn::FnArg) -> syn::Result<SelfType> {
-            None => FnType::Fn(parse_receiver(
-                "static method needs #[staticmethod] attribute",
-            )?),
         };
 
-        // parse rest of arguments
+        Ok(text_signature)
+    }
+
+    fn parse_arguments(
+        // inputs: &'a mut [syn::FnArg],
+        inputs_iter: impl Iterator<Item = &'a mut syn::FnArg>,
+    ) -> syn::Result<Vec<FnArg<'a>>> {
+        let mut arguments = vec![];
         for input in inputs_iter {
             match input {
                 syn::FnArg::Receiver(recv) => {
                     bail_spanned!(recv.span() => "unexpected receiver for method")
-                }
-                syn::FnArg::Typed(syn::PatType { pat, ty, .. }) => {
-                    let (ident, by_ref, mutability) = match &**pat {
+                } // checked in parse_fn_type
+                syn::FnArg::Typed(cap) => {
+                    let arg_attrs = PyFunctionArgAttrs::from_attrs(&mut cap.attrs)?;
+                    let (ident, by_ref, mutability) = match *cap.pat {
                         syn::Pat::Ident(syn::PatIdent {
-                            ident,
-                            by_ref,
-                            mutability,
+                            ref ident,
+                            ref by_ref,
+                            ref mutability,
                             ..
                         }) => (ident, by_ref, mutability),
-                        _ => bail_spanned!(pat.span() => "unsupported argument"),
+                        _ => bail_spanned!(cap.pat.span() => "unsupported argument"),
                     };
 
                     arguments.push(FnArg {
                         name: ident,
                         by_ref,
                         mutability,
-                        ty,
-                        optional: utils::option_type_argument(ty),
-                        py: utils::is_python(ty),
+                        ty: cap.ty.deref(),
+                        optional: utils::option_type_argument(cap.ty.deref()),
+                        py: utils::is_python(cap.ty.deref()),
+                        attrs: arg_attrs,
                     });
                 }
             }
         }
 
-        let ty = get_return_info(&sig.output);
-        let python_name = python_name.as_ref().unwrap_or(name).unraw();
+        Ok(arguments)
+    }
 
-        let mut parse_erroneous_text_signature = |error_msg: &str| {
-            // try to parse anyway to give better error messages
-            if let Some(text_signature) =
-                utils::parse_text_signature_attrs(meth_attrs, &python_name)?
-            {
-                bail_spanned!(text_signature.span() => error_msg)
+    fn parse_fn_type(
+        sig: &syn::Signature,
+        fn_type_attr: Option<MethodTypeAttribute>,
+        python_name: &mut Option<syn::Ident>,
+    ) -> syn::Result<(FnType, bool)> {
+        let name = &sig.ident;
+        let mut inputs_iter = sig.inputs.iter().enumerate();
```

Comment:
It looks to me that this is only used once in a `let _ = inputs_iter.next()` for `#[classmethod]`, so my conclusion is that we don't need this at all any more and can throw it away?

```suggestion
```

Reply:
fixed, thanks!